proposals/NNNN-clock-instant-date-duration.md

```diff
@@ -0,0 +1,498 @@
+
+Similarly to how [CGFloat was offered a special case for conversion](https://github.com/apple/swift-evolution/blob/main/proposals/0307-allow-interchangeable-use-of-double-cgfloat-types.md), Duration should have a special conversion case to TimeInterval to aide in the ergonomics of making sure the types are approachable. This means that any API that currently takes a TimeInterval now can take a Duration, and any API that takes a duration can take a concrete TimeInterval value. Just as the `CGFloat` to `Double` conversion was not taken lightly - this also is not a small issue. Expressing durations as a `Double` not only is potentially lossy but also pollutes the potential namespace with perhaps dubious concepts like multiplying two `TimeInterval` variables together is perhaps not the most meaningful usage. `Duration` being structured means that the type can be opinionated in what types of conformances it has, and operations can be extended upon it without mucking with unrelated categories.
+
+Meaningful durations can always be expressed in terms of nanoseconds, either a duration before a reference point or after. They can be constructed from meaningful human measured (or machine measured precision) but should not account for any calendrical calculations (e.g. a measure of days, months or years distinctly need a calendar to be meaningful). Durations should able to be serialized, compared, and stored as keys, but also should be able to be added and subtracted (and zero is meaningful). They are distinctly NOT `Numeric` due to the aforementioned issue with regards to multiplying two `TimeInterval` variables. That being said there is utility for ad-hoc division and multiplication to calculate back-offs.
+
+```swift
+public struct Duration: Sendable {
+  public var nanoseconds: Int64
+  
+  public init<T: BinaryInteger>(nanoseconds: T)
+}
+
+extension Duration {
+  public static func seconds<T: BinaryInteger>(_ seconds: T) -> Duration
+  public static func seconds(_ seconds: Double) -> Duration
+  public static func milliseconds<T: BinaryInteger>(_ milliseconds: T) -> Duration
+  public static func milliseconds(_ milliseconds: Double) -> Duration
+  public static func microseconds<T: BinaryInteger>(_ microseconds: T) -> Duration
+  public static func microseconds(_ microseconds: Double) -> Duration
+  public static func nanoseconds<T: BinaryInteger>(_ value: T) -> Duration 
+}
+
+extension Duration: Codable { }
+extension Duration: Hashable { }
+extension Duration: Equatable { }
+extension Duration: Comparable { }
+extension Duration: AdditiveArithmetic { }
+
+extension Duration {
+  public static func / (_ lhs: Duration, _ rhs: Double) -> Duration
+  public static func /= (_ lhs: inout Duration, _ rhs: Double)
+  public static func / <T: BinaryInteger>(_ lhs: Duration, _ rhs: T) -> Duration
+  public static func /= <T: BinaryInteger>(_ lhs: inout Duration, _ rhs: T)
```

Comment:
Curious: why do we have a generic one for BinaryInteger but not a one for BinaryFloatingPoint, aka

`public static func / <T: BinaryFloatingPoint>(_ lhs: Duration, _ rhs: T) -> Duration`

Reply:
I considered that, but the folks from the numerics team reminded me that the only used time type is Double (and conversions if needed to that are pretty simple)